File: ios/TapScore/Features/Friends/FriendProfileModel.swift
import Foundation

/// The two ways a friend-profile read can be refused, each a real state rather
/// than an error: the server answers 403 when the mutual edge is gone (or was
/// never there) and 404 when the player does not exist or was deleted. Neither
/// is retryable, so neither renders as an alert — the screen shows a calm
/// full-screen message and the back button is the way out.
enum FriendProfileUnavailability: Equatable, Sendable {
    /// The friendship was withdrawn while the viewer was looking (or the link
    /// was stale). Deliberately vague copy: "Anna removed you" would report one
    /// player's action to another.
    case forbidden
    /// The player id resolves to nothing — deleted account, dead link.
    case notFound

    var title: String {
        switch self {
        case .forbidden: return "Profile not available"
        case .notFound: return "Player not found"
        }
    }

    var message: String {
        switch self {
        case .forbidden: return "This profile is no longer shared with you."
        case .notFound: return "This player doesn't exist anymore."
        }
    }

    var systemImage: String {
        switch self {
        case .forbidden: return "eye.slash"
        case .notFound: return "person.slash"
        }
    }
}

/// Pure presentation rules for the friend-profile surfaces — outside SwiftUI
/// for the same reason `FriendsActivityModel` is: these are the lines a
/// reviewer argues about (what a round row says about someone else's score,
/// when a list admits it is not the whole story), and a view body is not a
/// place a rule can be asserted.
enum FriendProfileModel {
    /// Maps a thrown transport error to a refusal state, or nil when it is an
    /// ordinary failure (network, 500) that should render as a retryable error.
    /// One implementation, shared by all three stores, so the profile card and
    /// the lists cannot disagree about what a 403 means.
    static func unavailability(for error: any Error) -> FriendProfileUnavailability? {
        switch error {
        case APIError.server(403, _): return .forbidden
        case APIError.server(404, _): return .notFound
        default: return nil
        }
    }

    /// A round row's title: the organizer's name for the occasion, falling back
    /// to where it was played. "Round" only when the payload carries neither.
    static func title(_ entry: FriendProfileRoundEntry) -> String {
        if let name = entry.name, !name.isEmpty { return name }
        if let course = entry.courseName, !course.isEmpty { return course }
        return "Round"
    }

    /// The subject's progress in words, without pretending progress that has
    /// not happened:
    ///
    /// - a `not_started` round says so and shows no scoreline;
    /// - an active round with no scored hole reads as teeing off, mirroring
    ///   `FriendsActivityModel.progress` — "Thru 0 · E" would be a scoreline
    ///   they have not played;
    /// - `scoreToPar` is null before a first scored hole, so the score half is
    ///   simply absent rather than dashed.
    static func progress(_ entry: FriendProfileRoundEntry) -> String {
        let holes = Int(entry.holesPlayed)
        switch entry.status {
        case .notStarted:
            return "Not started"
        case .active:
            guard holes > 0 else { return "Teeing off" }
            guard let toPar = entry.scoreToPar else { return "Thru \(holes)" }
            return "Thru \(holes) · \(FriendsActivityModel.scoreToPar(Int(toPar)))"
        case .complete:
            guard holes > 0 else { return "Finished" }
            let played = holes < Int(entry.holeCount) ? "Thru \(holes)" : "Finished"
            guard let toPar = entry.scoreToPar else { return played }
            return "\(played) · \(FriendsActivityModel.scoreToPar(Int(toPar)))"
        }
    }

    /// A round's calendar date the way the rest of the app writes one —
    /// "12 May 2026", matching `RoundListView.displayDate` — falling back to
    /// the raw string when it does not parse (`rounds.date` is caller-supplied
    /// and only non-emptiness is guaranteed). Parsed with a UTC-pinned day
    /// formatter because the value is a zone-free calendar date: interpreting
    /// it in the device zone would shift it by a day for some users.
    static func displayDate(_ raw: String) -> String {
        guard let parsed = dayFormatter.date(from: raw) else { return raw }
        return parsed.formatted(.dateTime.day().month(.abbreviated).year())
    }

    nonisolated(unsafe) private static let dayFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.calendar = Calendar(identifier: .gregorian)
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone(identifier: "UTC")
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter
    }()

    /// One course row's fact line: "3 rounds · last played 12 May 2026".
    static func courseLine(_ course: FriendProfileCourseEntry) -> String {
        let rounds = Int(course.roundsPlayed)
        let played = rounds == 1 ? "1 round" : "\(rounds) rounds"
        return "\(played) · last played \(displayDate(course.lastPlayedAt))"
    }

    /// "4 courses played" for the profile's courses card.
    static func coursesSummary(_ total: Int) -> String {
        total == 1 ? "1 course played" : "\(total) courses played"
    }

    /// The header's identity line — "Hcp 9.0 · Linköpings GK" — with absent
    /// halves omitted, never dashed. Nil when both are absent, so the header
    /// can drop the line entirely.
    static func identityLine(handicapIndex: Double?, homeClubName: String?) -> String? {
        var parts: [String] = []
        if let handicap = handicapIndex {
            parts.append("Hcp \(FriendListModel.handicap(handicap))")
        }
        if let club = homeClubName, !club.isEmpty {
            parts.append(club)
        }
        return parts.isEmpty ? nil : parts.joined(separator: " · ")
    }

    /// The live line under the header — presence is the feed's call (activity
    /// recency), this only words it. Progress rides along when there is any;
    /// a live friend with no scored hole yet is teeing off, same as the strip.
    static func presenceLine(_ presence: FriendProfilePresence) -> String {
        guard presence.holesPlayed > 0 else { return "On the course now · Teeing off" }
        guard let toPar = presence.scoreToPar else {
            return "On the course now · Thru \(presence.holesPlayed)"
        }
        return "On the course now · Thru \(presence.holesPlayed) · \(FriendsActivityModel.scoreToPar(toPar))"
    }

    /// Stitches the next page onto what is already shown, dropping any round
    /// the list already carries. The keyset cursor cannot duplicate rows on a
    /// stable server, but a round created between two page reads can shift the
    /// window — a duplicate row would crash `ForEach` on its id, so the merge
    /// is defensive here rather than hopeful there.
    static func merge(
        _ existing: [FriendProfileRoundEntry],
        _ page: [FriendProfileRoundEntry]
    ) -> [FriendProfileRoundEntry] {
        let known = Set(existing.map(\.roundId))
        return existing + page.filter { !known.contains($0.roundId) }
    }
}

File: src/friends/friend-profile-model.ts
// Pure presentation rules for the friend-profile surfaces — the web twin of
// iOS `FriendProfileModel`, and the same product: same cards, same copy, same
// rules. These are the lines a reviewer argues about (what a round row says
// about someone else's score, when a list admits it is not the whole story,
// what a 403 means), so they live outside any component and are asserted in
// tests. No DOM, no fetch.

import { ApiError } from '@basics/core/client/api-error';
import type {
    FriendProfileCourseEntry,
    FriendProfileRoundEntry,
    FriendProfileRoundPage,
} from '../api/friend-profile.gen';
import { scoreToParText } from './friends-activity-model';

/**
 * The two ways a friend-profile (or spectate) read can be refused, each a real
 * state rather than an error: 403 when the mutual edge is gone (or was never
 * there), 404 when the target does not exist. Neither is retryable, so neither
 * renders as a toast or a retry loop — the screen shows a calm full-page
 * message and back is the way out.
 */
export type FriendProfileUnavailability = 'forbidden' | 'not_found';

/**
 * Maps a thrown transport error to a refusal state, or null when it is an
 * ordinary failure (network, 500) that should render as a retryable error.
 * One implementation shared by the profile, the lists and the spectate view,
 * so no two surfaces can disagree about what a 403 means.
 */
export function unavailability(err: unknown): FriendProfileUnavailability | null {
    if (err instanceof ApiError) {
        if (err.status === 403) return 'forbidden';
        if (err.status === 404) return 'not_found';
    }
    return null;
}

/**
 * The refusal copy. Deliberately vague on 'forbidden': "Anna removed you"
 * would report one player's action to another.
 */
export const UNAVAILABILITY_COPY: Record<
    FriendProfileUnavailability,
    { title: string; message: string }
> = {
    forbidden: {
        title: 'Profile not available',
        message: 'This profile is no longer shared with you.',
    },
    not_found: {
        title: 'Player not found',
        message: "This player doesn't exist anymore.",
    },
};

/** A round row's title: the organizer's name for the occasion, falling back to
 *  where it was played. "Round" only when the payload carries neither. */
export function roundTitle(
    entry: Pick<FriendProfileRoundEntry, 'name' | 'courseName'>,
): string {
    const name = (entry.name ?? '').trim();
    if (name) return name;
    const course = (entry.courseName ?? '').trim();
    return course || 'Round';
}

/**
 * The row's subtitle: the (formatted) date, plus the course when the title is
 * the organizer's own name for the round — otherwise the course IS the title
 * and repeating it says nothing. `formatDate` is injected so this module stays
 * locale-free; the components pass `formatRowDate`.
 */
export function roundSubtitle(
    entry: Pick<FriendProfileRoundEntry, 'name' | 'courseName' | 'date'>,
    formatDate: (date: string) => string,
): string {
    const date = formatDate(entry.date);
    const name = (entry.name ?? '').trim();
    const course = (entry.courseName ?? '').trim();
    if (name && course) return `${date} · ${course}`;
    return date;
}

/**
 * The subject's progress in words, without pretending progress that has not
 * happened — the same vocabulary as iOS:
 *
 * - a `not_started` round says so and shows no scoreline;
 * - an active round with no scored hole reads as teeing off, mirroring
 *   `friendProgress` — "Thru 0 · E" would be a scoreline they have not played;
 * - `scoreToPar` is null before a first scored hole, so the score half is
 *   simply absent rather than dashed.
 */
export function roundProgress(
    entry: Pick<FriendProfileRoundEntry, 'status' | 'holesPlayed' | 'holeCount' | 'scoreToPar'>,
): string {
    const holes = entry.holesPlayed;
    switch (entry.status) {
        case 'not_started':
            return 'Not started';
        case 'active': {
            if (holes <= 0) return 'Teeing off';
            if (entry.scoreToPar === null) return `Thru ${holes}`;
            return `Thru ${holes} · ${scoreToParText(entry.scoreToPar)}`;
        }
        case 'complete': {
            if (holes <= 0) return 'Finished';
            const played = holes < entry.holeCount ? `Thru ${holes}` : 'Finished';
            if (entry.scoreToPar === null) return played;
            return `${played} · ${scoreToParText(entry.scoreToPar)}`;
        }
    }
}

/** The header's identity line — "Hcp 9.0 · Linköpings GK" — with absent halves
 *  omitted, never dashed. Null when both are absent, so the header drops the
 *  line entirely. */
export function identityLine(
    handicapIndex: number | null,
    homeClubName: string | null,
): string | null {
    const parts: string[] = [];
    if (handicapIndex !== null) parts.push(`Hcp ${handicapIndex.toFixed(1)}`);
    const club = (homeClubName ?? '').trim();
    if (club) parts.push(club);
    return parts.length > 0 ? parts.join(' · ') : null;
}

/** One course row's fact line: "3 rounds · last played 12 May 2026". */
export function courseLine(
    course: Pick<FriendProfileCourseEntry, 'roundsPlayed' | 'lastPlayedAt'>,
    formatDate: (date: string) => string,
): string {
    const played = course.roundsPlayed === 1 ? '1 round' : `${course.roundsPlayed} rounds`;
    return `${played} · last played ${formatDate(course.lastPlayedAt)}`;
}

/** "4 courses played" for the profile's courses card. */
export function coursesSummary(total: number): string {
    return total === 1 ? '1 course played' : `${total} courses played`;
}

// --- Paged round list -------------------------------------------------------

/**
 * The see-all list's whole pagination state. The cursor is OPAQUE: it is
 * carried back to the server verbatim and never constructed or inspected here.
 * `hasMore` is the only stop condition — the list's length is NEVER compared
 * against the profile card's `roundsTotal`, because the two disagree by design
 * (private and link rounds count there and are absent here).
 */
export interface RoundListState {
    rounds: FriendProfileRoundEntry[];
    nextCursor: string | null;
    hasMore: boolean;
}

export const EMPTY_ROUND_LIST: RoundListState = { rounds: [], nextCursor: null, hasMore: false };

/**
 * Stitches the next page onto what is already shown, dropping any round the
 * list already carries. The keyset cursor cannot duplicate rows on a stable
 * server, but a round created between two page reads can shift the window — a
 * duplicate row would break the `$each` key, so the merge is defensive here
 * rather than hopeful there. The page's cursor and `hasMore` ride through
 * verbatim.
 */
export function stitchPage(state: RoundListState, page: FriendProfileRoundPage): RoundListState {
    const known = new Set(state.rounds.map((r) => r.roundId));
    return {
        rounds: [...state.rounds, ...page.rounds.filter((r) => !known.has(r.roundId))],
        nextCursor: page.nextCursor,
        hasMore: page.hasMore,
    };
}

/** Whether another page may be asked for — `hasMore` alone decides, with the
 *  cursor's presence as a belt-and-braces guard against a malformed page. */
export function canLoadMore(state: Pick<RoundListState, 'hasMore' | 'nextCursor'>): boolean {
    return state.hasMore && state.nextCursor !== null;
}
